engine/src/main/java/org/hibernate/validator/internal/engine/cascading/ValueExtractorManager.java

```diff
@@ -9,70 +9,119 @@
 import java.lang.reflect.TypeVariable;
 import java.security.AccessController;
 import java.security.PrivilegedAction;
-import java.util.ArrayList;
 import java.util.Collections;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import javax.validation.ValidationException;
 import javax.validation.valueextraction.ValueExtractor;
 
 import org.hibernate.validator.internal.util.TypeHelper;
 import org.hibernate.validator.internal.util.TypeVariableBindings;
 import org.hibernate.validator.internal.util.privilegedactions.LoadClass;
+import org.hibernate.validator.internal.util.stereotypes.Immutable;
 
 /**
  * @author Gunnar Morling
  * @author Guillaume Smet
  */
 public class ValueExtractorManager {
 
-	private final List<ValueExtractorDescriptor> valueExtractors;
+	@Immutable
+	public static final Set<ValueExtractor<?>> SPEC_DEFINED_EXTRACTORS = Stream.of(
+			ByteArrayValueExtractor.DESCRIPTOR,
+			ShortArrayValueExtractor.DESCRIPTOR,
+			IntArrayValueExtractor.DESCRIPTOR,
+			LongArrayValueExtractor.DESCRIPTOR,
+			FloatArrayValueExtractor.DESCRIPTOR,
+			DoubleArrayValueExtractor.DESCRIPTOR,
+			CharArrayValueExtractor.DESCRIPTOR,
+			BooleanArrayValueExtractor.DESCRIPTOR,
+			ObjectArrayValueExtractor.DESCRIPTOR,
 
-	public ValueExtractorManager(Iterable<ValueExtractor<?>> externalExtractors) {
-		List<ValueExtractorDescriptor> tmpValueExtractors = new ArrayList<>();
+			LegacyListValueExtractor.DESCRIPTOR,
+			ListValueExtractor.DESCRIPTOR,
+
+			LegacyMapValueExtractor.DESCRIPTOR,
+			MapValueExtractor.DESCRIPTOR,
+			MapKeyExtractor.DESCRIPTOR,
+
+			LegacyIterableValueExtractor.DESCRIPTOR,
+			IterableValueExtractor.DESCRIPTOR,
+
+			OptionalValueExtractor.DESCRIPTOR,
+			ObjectValueExtractor.DESCRIPTOR
+		)
+		.map( ValueExtractorDescriptor::getValueExtractor )
+		.collect( Collectors.collectingAndThen( Collectors.toSet(), Collections::unmodifiableSet ) );
+
+	@Immutable
+	private final Map<ValueExtractorDescriptor.Key, ValueExtractorDescriptor> valueExtractors;
+
+	public ValueExtractorManager(Set<ValueExtractor<?>> externalExtractors) {
+		LinkedHashMap<ValueExtractorDescriptor.Key, ValueExtractorDescriptor> tmpValueExtractors = new LinkedHashMap<>();
 
 		for ( ValueExtractor<?> valueExtractor : externalExtractors ) {
-			tmpValueExtractors.add( new ValueExtractorDescriptor( valueExtractor ) );
+			ValueExtractorDescriptor descriptor = new ValueExtractorDescriptor( valueExtractor );
+			tmpValueExtractors.put( descriptor.getKey(), descriptor );
```

Comment:
I wonder if it wouldn't be better to add the user defined VE after the default ones, thus providing a way to override the default ones.